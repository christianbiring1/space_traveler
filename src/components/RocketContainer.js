import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import toast, { Toaster } from 'react-hot-toast';
import { cancelReservation, reserveRocket } from '../redux/rockets/rockets';
import '../styles/rocketContainer.scss';

function RocketContainer({
  id,
  name,
  description,
  image,
  reserved,
}) {
  const dispatch = useDispatch();

  const toastDisplay = (message) => {
    toast.success(message, {
      style: {
        border: '1px solid blue',
        padding: '16px',
        color: 'blue',
      },
      iconTheme: {
        primary: 'blue',
        secondary: '#FFFAEE',
      },
    });
  };

  const reserveRocketHandler = (id, name) => {
    dispatch(reserveRocket(id));
    toastDisplay(`${name} has been Reserved`);
  };

  const cancelReservationHandler = (id, name) => {
    dispatch(cancelReservation(id));
    toastDisplay(`${name} reservation has been Cancelled`);
  };

  const ReservedBadge = () => <b className="badgeIcon">Reserved</b>;

  return (
    <div className="rocketContainer" data-testid="rocketContainer">
      <Toaster />
      <div className="imageContainer">
        <img src={image} alt="rocket" className="image" />
      </div>
      <div className="rocketDetails">
        <h3 className="rocketName">{name}</h3>
        <p className="rocketDescription">
          {reserved
            ? [<ReservedBadge key={1} />, ' ', description]
            : description}
        </p>
        {reserved
          ? <button type="button" className="cancelReservationButton" onClick={() => cancelReservationHandler(id, name)}>Cancel Reservation</button>
          : <button type="button" className="reserveRocketocketButton" onClick={() => reserveRocketHandler(id, name)}>Reserve Rocket</button>}
      </div>
    </div>
  );
}

RocketContainer.propTypes = {
  id: PropTypes.number,
  name: PropTypes.string,
  reserved: PropTypes.bool,
  image: PropTypes.string,
  description: PropTypes.string,
};

RocketContainer.defaultProps = {
  id: 1,
  name: 'Rocket',
  reserved: false,
  image: 'https://image.com',
  description: 'Lorem ipsum dolor sit amet.',
};

export default RocketContainer;
