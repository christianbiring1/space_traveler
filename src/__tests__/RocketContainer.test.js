import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import RocketContainer from '../components/RocketContainer';

describe('Rocket page tests', () => {
  const rocket = {
    id: 1,
    rocket_name: 'Rocket',
    description: 'Lorem ipsum dolor sit amet',
    flickr_image: 'https://img.example.com',
    reserved: false,
  };
  it('Page renders as expected', () => {
    const component = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <RocketContainer
            key={rocket.id}
            id={rocket.id}
            name={rocket.rocket_name}
            description={rocket.description}
            image={rocket.flickr_image}
            reserved={rocket.reserved}
          />
        </BrowserRouter>
      </Provider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Page is visible to the user', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <BrowserRouter>
          <RocketContainer />
        </BrowserRouter>
      </Provider>,
    );
    expect(getByTestId('rocketContainer')).toBeInTheDocument();
  });
});
