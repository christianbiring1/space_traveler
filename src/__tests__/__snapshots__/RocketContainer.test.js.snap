// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rocket page tests Page renders as expected 1`] = `
<div
  className="rocketContainer"
  data-testid="rocketContainer"
>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "bottom": 16,
        "left": 16,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 16,
        "top": 16,
        "zIndex": 9999,
      }
    }
  />
  <div
    className="imageContainer"
  >
    <img
      alt="rocket"
      className="image"
      src="https://img.example.com"
    />
  </div>
  <div
    className="rocketDetails"
  >
    <h3
      className="rocketName"
    >
      Rocket
    </h3>
    <p
      className="rocketDescription"
    >
      Lorem ipsum dolor sit amet
    </p>
    <button
      className="reserveRocketocketButton"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
